testnet/stacks-node/src/event_dispatcher.rs

```diff
@@ -102,10 +124,59 @@ pub struct MinedNakamotoBlockEvent {
     pub stacks_height: u64,
     pub block_size: u64,
     pub cost: ExecutionCost,
+    pub miner_signature: MessageSignature,
+    pub signer_signature_hash: Sha512Trunc256Sum,
     pub tx_events: Vec<TransactionEvent>,
     pub signer_bitvec: String,
 }
 
+impl StackerDBChannel {
+    pub fn new() -> Self {
+        let (sender, recv_channel) = std::sync::mpsc::channel();
+        Self {
+            receiver: Mutex::new(Some(recv_channel)),
+            sender,
+        }
+    }
+
+    /// Return the receiver to the StackerDBChannel. This must be done before
+    ///  another interested thread can subscribe to events.
+    ///
+    /// The StackerDBChnnel's receiver is guarded with a Mutex, so that ownership can
+    /// be taken by different threads without unsafety.
+    pub fn replace_receiver(&self, receiver: Receiver<StackerDBChunksEvent>) {
```

Comment:
This feels brittle.  I realize that there should never be two miner threads, but this API requires the miner implementation to _perfectly_ take and replace the receiver on each code path it could possibly take during instantiation and tear-down, or risk leaving the system in a broken state.  Is there a way to instead have the miner _subscribe itself_ to the channel by registering its own singleton receiver (e.g. paired to its thread ID by this code), and have the subscription logic garbage-collect stale receivers?  Then, all the miner thread has to do is subscribe for events, without any thought to what happens afterwards.

Reply:
The SignerCoordinator, which obtains the receiver on instantiation, defines `Drop`. I think that the `take()` should be called closer to the end of the instantiation, so that there's no fallible code between it and the constructor.

The problem with having the signer coordinator push its receiver is that sender threads would need to get new send channels ever time a new signer coordinator was instantiated. The way its implemented here, once a thread has ownership of a sender, it doesn't need to refresh that view. It's certainly possible to do, but would involve a mutex on the event dispatching side (i.e., the sender).